File: go-arena/internal/game/spawner.go
package game

import "arena-server/internal/config"

// SpawnBotAt places a bot at a specific position and resets its combat state.
// If the chosen cell is blocked it searches outward for the nearest open cell.
func SpawnBotAt(bot *BotState, pos Vec2, grid *SpatialGrid, tickCount int) {
	// Snap position to the nearest grid cell centre. If the cell is a wall,
	// take the nearest open cell anywhere on the grid — a bounded search
	// (formerly radius 10) could give up inside thick caves walls and leave
	// the bot permanently embedded.
	if ActiveTerrain != nil {
		cell := ActiveTerrain.WorldToGrid(pos)
		if open, ok := ActiveTerrain.NearestOpenCell(cell, 0); ok {
			cell = open
		}
		pos = ActiveTerrain.GridToWorld(cell)
	}
	bot.Position = pos
	bot.LastValidPosition = pos
	bot.HP = bot.MaxHP
	bot.IsAlive = true
	bot.CooldownRemaining = 0

	bot.ActiveEffects = nil
	bot.DodgeCooldown = 0
	bot.InvulnTicks = 0
	bot.StunTicks = 0
	bot.ShieldAbsorb = 0

	bot.CurrentPath = nil
	bot.PathTarget = nil
	bot.MovementTrace = nil

	bot.GrappleCharges = config.C.GrappleChargesPerRound
	bot.GrappleCooldown = 0
	bot.BountyTokenBonus = 0
	bot.TeleportHazardGraceTicks = 0
	bot.TeleportTouchedPads = make(map[string]bool)

	bot.RoundLifeStartTick = tickCount
	bot.ResetDamageAttribution()

	grid.Insert(bot.BotID, bot.Position)
}

const killCreditTTLSeconds = 5

func killCreditTTLTicks() int {
	tickRate := config.C.TickRate
	if tickRate <= 0 {
		tickRate = 10
	}
	return killCreditTTLSeconds * tickRate
}

func recentDamageKiller(victim *BotState, bots map[string]*BotState, tickCount int) *BotState {
	if victim == nil || victim.LastDamagedBy == "" || victim.LastDamagedBy == victim.BotID {
		return nil
	}

	age := tickCount - victim.LastDamageTick
	if age < 0 || age > killCreditTTLTicks() {
		return nil
	}

	return bots[victim.LastDamagedBy]
}

// CheckDeaths finds all bots whose HP has reached zero, marks them dead,
// removes them from the spatial grid, and returns a DeathEvent for each.
func CheckDeaths(bots map[string]*BotState, grid *SpatialGrid, tickCount int) []DeathEvent {
	var events []DeathEvent

	for _, bot := range bots {
		if !bot.IsAlive || bot.HP > 0 {
			continue
		}

		bot.IsAlive = false
		bot.HP = 0
		grid.Remove(bot.BotID)

		death := DeathEvent{
			VictimID:    bot.BotID,
			VictimKills: bot.RoundKills,
		}
		if killer := recentDamageKiller(bot, bots, tickCount); killer != nil {
			death.KillerID = killer.BotID
			death.KillerName = killer.Name
			death.Weapon = bot.LastDamageSource
			death.Damage = bot.LastDamageAmount
		}
		events = append(events, death)

		bot.RoundDeaths++
	}

	return events
}

// No respawns - dead bots stay dead until next round.

File: server/game/spawner.py
"""Spawn, respawn, and death handling for the game engine."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from server.config import settings

if TYPE_CHECKING:
    from server.game.arena_map import ArenaMap
    from server.game.spatial import SpatialGrid
    from server.game.state import BotState

logger = logging.getLogger(__name__)


def spawn_bot(bot: BotState, arena: ArenaMap, grid: SpatialGrid) -> None:
    """Place a bot at a random position in the safe zone."""
    pos = arena.get_random_spawn_point()
    bot.position = pos
    bot.spawn_position = pos
    bot.is_alive = True
    bot.hp = bot.max_hp
    bot.cooldown_remaining = 0.0
    bot.respawn_timer = 0.0
    bot.active_effects.clear()
    bot.dodge_cooldown = 0
    bot.invuln_ticks = 0
    bot.stun_ticks = 0
    bot.shield_absorb = 0
    bot.current_path.clear()
    bot.path_target = None
    grid.insert(bot.bot_id, pos[0], pos[1])
    logger.info("Bot %s spawned at (%.1f, %.1f)", bot.name, pos[0], pos[1])


def check_deaths(
    bots: dict[str, BotState],
    grid: SpatialGrid,
    tick_count: int,
) -> list[dict]:
    """Check for dead bots (hp <= 0), emit death events, start respawn timers.

    Returns list of death event dicts.
    """
    events: list[dict] = []
    respawn_time = settings.combat.respawn_time

    for bot_id, bot in bots.items():
        if not bot.is_alive or bot.hp > 0:
            continue

        bot.is_alive = False
        bot.hp = 0
        bot.respawn_timer = float(respawn_time)
        bot.current_path.clear()
        bot.path_target = None
        grid.remove(bot_id)

        events.append({
            "type": "death",
            "bot_id": bot_id,
            "bot_name": bot.name,
            "kills_this_life": bot.kill_streak,
            "tick": tick_count,
        })

        bot.kill_streak = 0
        logger.info("Bot %s died at tick %d", bot.name, tick_count)

    return events


def process_respawns(
    bots: dict[str, BotState],
    arena: ArenaMap,
    grid: SpatialGrid,
    tick_rate: int,
) -> list[dict]:
    """Tick down respawn timers and respawn bots when ready.

    Returns list of respawn event dicts.
    """
    events: list[dict] = []
    dt = 1.0 / tick_rate

    for bot_id, bot in bots.items():
        if bot.is_alive or bot.respawn_timer <= 0:
            continue

        bot.respawn_timer -= dt
        if bot.respawn_timer <= 0:
            bot.respawn_timer = 0.0
            pos = arena.get_random_spawn_point()
            bot.position = pos
            bot.spawn_position = pos
            bot.is_alive = True
            bot.hp = bot.max_hp
            bot.cooldown_remaining = 0.0
            bot.active_effects.clear()
            bot.dodge_cooldown = 0
            bot.invuln_ticks = 0
            bot.stun_ticks = 0
            bot.shield_absorb = 0
            bot.current_path.clear()
            bot.path_target = None
            bot.last_damaged_by = None
            grid.insert(bot_id, pos[0], pos[1])

            events.append({
                "type": "respawn",
                "bot_id": bot_id,
                "position": pos,
                "hp": bot.hp,
            })
            logger.info("Bot %s respawned at (%.1f, %.1f)", bot.name, pos[0], pos[1])

    return events
